### PR TITLE
[DeviceSanitizer] Support detecting out-of-bounds error on DeviceGlobals

### PR DESCRIPTION
**Pull request by zhaomaosu**
UR part: https://github.com/oneapi-src/unified-runtime/pull/1354

### CODE REVIEW

File: llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp
Comment (AllanZyne) [inline llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp:2372]: How about using a more descriptive variable name for "beg"?

**Review by AllanZyne — APPROVED** [review]
LGTM. 

Please modify `UNIFIED_RUNTIME_REPO` and `UNIFIED_RUNTIME_TAG` in "llvm/sycl/plugins/unified_runtime/CMakeLists.txt" to enable integration test.

File: llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp
Comment (wenju-he) [inline llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp:2376]: will device globals in https://github.com/intel/llvm/blob/626c121cf2594df368636286585628d1189b87aa/libdevice/sanitizer_utils.cpp#L28-L35 be instrumented here as well?

Comment (jinge90) [reply llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp:2376]: Device global in libdevice shouldn't be affected, the Asan Instrument Pass will only go user's device image in device compilation phase. Does this align with your expectation? @zhaomaosu

**Comment by jinge90** [comment]
Hi, @zhaomaosu 
Does this patch support following scenario:
clang++ -fsycl -fsanitize=address -c user_code1.cpp -o user1.o
clang++ -fsycl -fsanitize=address -c user_code2.cpp -o user2.o
clang++ -fsycl user1.o user2.o -o user.exe
Both user_code1.cpp and user_code2.cpp include device globals.
Thanks.

File: llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp
Comment (zhaomaosu) [reply llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp:2376]: Yeah, device globals in libdevice will not be affected since they don't have any properties.

Comment (wenju-he) [reply llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp]: add a comment that hasDeviceImageScopeProperty is used for excluding device globals in libdevice? It isn't very intuitive why hasDeviceImageScopeProperty is called here.

File: llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp
Comment (wenju-he) [inline llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp]: why non-DeviceGlobalVariable global variables aren't instrumented here?

**Comment by jinge90** [comment]
> Hi, @zhaomaosu Does this patch support following scenario: clang++ -fsycl -fsanitize=address -c user_code1.cpp -o user1.o clang++ -fsycl -fsanitize=address -c user_code2.cpp -o user2.o clang++ -fsycl user1.o user2.o -o user.exe Both user_code1.cpp and user_code2.cpp include device globals. Thanks.

Hi, @zhaomaosu 
Could you add some tests to cover this scenario?
Thanks.

File: llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp
Comment (zhaomaosu) [reply llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp]: done

File: llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp
Comment (zhaomaosu) [reply llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp]: Now, we only care about global variables that are visible to the user. So, we only instrument device global variables. In the future, we can easily extend to instrument other user visible global variables.

**Comment by zhaomaosu** [comment]
> Hi, @zhaomaosu Could you add some tests to cover this scenario? Thanks.

Ok, I'll add tests for it.

**Comment by zhaomaosu** [comment]
> Hi, @zhaomaosu Could you add some tests to cover this scenario? Thanks.

Done

File: llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp
Comment (AllanZyne) [reply llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp]: > add a comment that hasDeviceImageScopeProperty is used for excluding device globals in libdevice?

hasDeviceImageScopeProperty is not used for excuding device globals in libdevice.
ASanPass only runs when user kernel code is compiling, and after it is compiled, it will link with libdevice.
So ASanPass doesn't modify libdevice's code.

Comment (wenju-he) [reply llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp]: ok. my comment is just based on the previous comment.
Then why is hasDeviceImageScopeProperty called here?

Comment (AllanZyne) [reply llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp]: > ok. my comment is just based on the previous comment. Then why is hasDeviceImageScopeProperty called here?

The device globals without "device_image_scope" property is implemented by the Device USM by the SYCL runtime, so it's automatically supported by us.
The device globals with "device_image_scope" property is more like the true global variable in one module.

File: llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp
Comment (wenju-he) [reply llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp]: at the moment I think global variables can come from openmp or third party library like in fpga-emulator case. It is fine to solve them in the next step.

File: llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp
Comment (wenju-he) [reply llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp]: a comment explaining what device globals are excluded by the function would more intuitive. But it is fine if you think it is unnecessary.

Comment (zhaomaosu) [reply llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp]: Thanks, added a comment for this. BTW, I have moved the instrument code to sycl-post-link tool.

File: llvm/tools/sycl-post-link/sycl-post-link.cpp
Comment (AllanZyne) [inline llvm/tools/sycl-post-link/sycl-post-link.cpp:891]: If this checking method is stable, can you help to change "isModuleUsingAsan" function, and use it here?

File: llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp
Comment (AllanZyne) [reply llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp]: Why did you move device global instrumentation in sycl-post-link?

Comment (zhaomaosu) [reply llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp]: If there are more than one user source files, since we will add extra global variables to each source file, multi definition errors will be reported by llvm-link tool.

File: llvm/tools/sycl-post-link/sycl-post-link.cpp
Comment (zhaomaosu) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp]: Done

File: llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp
Comment (wenju-he) [reply llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp]: can we still instrument in instrumentation pass and adds AsanDeviceGlobalMetadata with appending linkage. Than in sycl-post-link we only need to create __AsanDeviceGlobalCount

Comment (wenju-he) [reply llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp]: please also check if clGetDeviceGlobalVariablePointerINTEL can get ride of __AsanDeviceGlobalCount

Comment (zhaomaosu) [reply llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp]: Discussed with Wenju offline, I'll create another PR to implement new UR API to get device global pointer. Before that, we will keep current solution. When the new API is implemented, I'll move this instrument part back to AddressSanitizer pass.

**Review by wenju-he — APPROVED** [review]
LGTM

**Comment by zhaomaosu** [comment]
@intel/llvm-reviewers-runtime , @intel/dpcpp-tools-reviewers could you please help review this PR? Thanks.

**Review by steffenlarsen — CHANGES_REQUESTED** [review]
Great addition! I'm blocking it to avoid merging the redirected UR Cmake changes, but as soon as it is updated you can have my approval. 👍

File: sycl/plugins/unified_runtime/CMakeLists.txt
Comment (yingcong-wu) [inline sycl/plugins/unified_runtime/CMakeLists.txt:59]: Wrong repo!

**Comment by zhaomaosu** [comment]
Hi @steffenlarsen, I noticed there are some failures in GEN12 tests, but I think my changes will not affect it. Do you know if these are known failures?

**Comment by steffenlarsen** [comment]
The Gen12 failures are reported in https://github.com/intel/llvm/issues/12797 and https://github.com/intel/llvm/issues/12798 so they are completely unrelated to these changes.

**Comment by zhaomaosu** [comment]
Hi @steffenlarsen, the PR in UR repo has been merged. I updated the UR tag. May I get your approval?

**Review by steffenlarsen — APPROVED** [review]
LGTM!

**Comment by steffenlarsen** [comment]
Review still required from @intel/unified-runtime-reviewers & @intel/dpcpp-tools-reviewers. Friendly ping.

**Review by aarongreig — APPROVED** [review]

File: llvm/tools/sycl-post-link/sycl-post-link.cpp
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/sycl-post-link.cpp:895]:
```suggestion
  constexpr uint64_t MinRZ = 32;
```

Nit, for uniformity

File: llvm/tools/sycl-post-link/sycl-post-link.cpp
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/sycl-post-link.cpp:898]:
```suggestion
      Type::getIntNTy(M.getContext(), DL.getPointerSizeInBits());
```

File: llvm/tools/sycl-post-link/sycl-post-link.cpp
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/sycl-post-link.cpp:897]:
```suggestion
  Type *IntTy =
```

Or maybe `SizeTTy`. That type is not a pointer, so `ptr` is definitely confusing here

File: llvm/tools/sycl-post-link/sycl-post-link.cpp
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/sycl-post-link.cpp:910]: This should be turned into an early exit instead

File: llvm/tools/sycl-post-link/sycl-post-link.cpp
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/sycl-post-link.cpp:931]: Formatting seems to be broken

File: llvm/tools/sycl-post-link/sycl-post-link.cpp
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/sycl-post-link.cpp:922]: Am I right that `(SizeInBytes / MinRZ / 4) * MinRZ` is written this way on purpose, to ensure certain rounding during calculation? If so, a comment would be welcome about that.

File: llvm/tools/sycl-post-link/sycl-post-link.cpp
Comment (zhaomaosu) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp]: Done

File: llvm/tools/sycl-post-link/sycl-post-link.cpp
Comment (zhaomaosu) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp]: Done

File: llvm/tools/sycl-post-link/sycl-post-link.cpp
Comment (zhaomaosu) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp]: Changed to IntTy

File: llvm/tools/sycl-post-link/sycl-post-link.cpp
Comment (zhaomaosu) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp]: Done

File: llvm/tools/sycl-post-link/sycl-post-link.cpp
Comment (zhaomaosu) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp]: Re-run clang-format

File: llvm/tools/sycl-post-link/sycl-post-link.cpp
Comment (zhaomaosu) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp]: This code is coming from https://github.com/intel/llvm/blob/sycl/llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp#L2767.
I traced back the git histories. It seems that this code initially started out like this. So, I also don't know if it's written this way for some reason. Sorry for this.
But we do this instrumentation for device global in sycl-post-link tool temporarily. When unified runtime implemented the API urProgramGetGlobalVariablePointer, I'll move this instrumentation code back to AddressSanitizer pass.

File: llvm/tools/sycl-post-link/sycl-post-link.cpp
Comment (maksimsab) [inline llvm/tools/sycl-post-link/sycl-post-link.cpp:887]: I think it should be a Pass rather than a simple function.

**Review by maksimsab — CHANGES_REQUESTED** [review]

File: llvm/tools/sycl-post-link/sycl-post-link.cpp
Comment (zhaomaosu) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp:887]: Hi @maksimsab, we just do this instrumentation in sycl-post-link temporarily. Once UR PR https://github.com/oneapi-src/unified-runtime/pull/1255 merged, I'll move this code back to AddressSanitizer pass. If you still think I need to write a pass to this, it's also ok for me. Do you think it's necessary to write a pass?

Comment (zhaomaosu) [reply llvm/tools/sycl-post-link/sycl-post-link.cpp]: Done, write a pass for this.

**Comment by zhaomaosu** [comment]
@maksimsab, any other comments?

**Comment by kbenzie** [comment]
@maksimsab if the changes you requested as sufficient please approve the PR.

@intel/dpcpp-esimd-reviewers please review this ASAP, this PR is blocking any progress of merging other UR changes.

This is affecting multiple teams. Any help to expedite the approval and merging of this PR would be greatly appriciated.

**Comment by maksimsab** [comment]
I am OK to let this PR go in order to unblock other teams.
Currently, the new Pass has 0 test coverage. @zhaomaosu  Are you ok to address this in the follow-up PR?

**Review by maksimsab — APPROVED** [review]
LGTM

**Comment by zhaomaosu** [comment]
> I am OK to let this PR go in order to unblock other teams. Currently, the new Pass has 0 test coverage. @zhaomaosu Are you ok to address this in the follow-up PR?

Sure, I'll do this later. Thank you.

**Comment by zhaomaosu** [comment]
Can anyone help merge this PR? Thanks a lot.

**Review by sarnex — APPROVED** [review]
esimd lgtm, leaving tools for others

**Pull request merged**
